<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="tantale">
<title>1. Mining TALE sequences in genomes with tantale • tantale</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="1. Mining TALE sequences in genomes with tantale">
<meta property="og:description" content="tantale">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">tantale</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.9550</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/1_tale_mining.html">1. Mining TALE sequences in genomes with tantale</a>
    <a class="dropdown-item" href="../articles/2_tale_classification.html">2. Classifying TALE sequences from genomes with tantale</a>
    <a class="dropdown-item" href="../articles/3_tale_msa.html">3. Multiple alignment of TALE sequences with tantale</a>
    <a class="dropdown-item" href="../articles/4_tale_target_prediction.html">4. TALE targets predictions with tantale</a>
    <a class="dropdown-item" href="../articles/p1_distalr.html">A comparison between `runDistal` and `distalr` implementations</a>
    <a class="dropdown-item" href="../articles/p2_multiple_alignments.html">TALEs domains sequences alignment with MAFFT</a>
    <a class="dropdown-item" href="../articles/p3_tantale_objects.html">An overview of the various objects related to TALEs in tantale</a>
  </div>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/scunnac/tantale/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">



<script src="1_tale_mining_files/accessible-code-block-0.0.1/empty-anchor.js"></script><div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>1. Mining TALE sequences in genomes with tantale</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/scunnac/tantale/blob/HEAD/../../../vignettes/1_tale_mining.Rmd" class="external-link"><code>../../../vignettes/1_tale_mining.Rmd</code></a></small>
      <div class="d-none name"><code>1_tale_mining.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://scunnac.github.io/tantale">tantale</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://ggplot2.tidyverse.org" class="external-link">ggplot2</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://tidyverse.tidyverse.org" class="external-link">tidyverse</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://bioconductor.org/packages/Biostrings" class="external-link">Biostrings</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va">parallel</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://magrittr.tidyverse.org" class="external-link">magrittr</a></span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/rstudio/DT" class="external-link">DT</a></span><span class="op">)</span></span></code></pre></div>
<p>In the first four articles, we illustrate the various major approaches employed in a typical study of TALE diversity and function from a set of bacterial genome sequences.</p>
<p>This first article covers all the steps in a TALE mining pipeline from “accurate” or “error-prone” Xoo genome sequences using functions in package <code>tantale</code>.<br>
Genomes used for this vignette are MAI1, BAI3, BAI3-1-1 (an error-prone genome with <em>talC</em> deleted) and PXO86, an outgroup Asian strain.<br></p>
<p>Get the fasta files:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mai1_fa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"MAI1.fa"</span>, package <span class="op">=</span> <span class="st">"tantale"</span>, mustWork <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">bai3_fa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"BAI3.fa"</span>, package <span class="op">=</span> <span class="st">"tantale"</span>, mustWork <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">bai311_fa</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"BAI3-1-1.fa"</span>, package <span class="op">=</span> <span class="st">"tantale"</span>, mustWork <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span><span class="va">pxo86</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"PXO86.fa"</span>, package <span class="op">=</span> <span class="st">"tantale"</span>, mustWork <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<p>Output dir to save all the results of this vignette:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">outdir</span> <span class="op">&lt;-</span> <span class="fu">fs</span><span class="fu">::</span><span class="fu"><a href="https://fs.r-lib.org/reference/create.html" class="external-link">dir_create</a></span><span class="op">(</span><span class="st">"~/TEMP/test_tantale"</span><span class="op">)</span> <span class="co"># tempdir(check = TRUE)</span></span>
<span><span class="va">telltale_outdir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"telltale"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/files2.html" class="external-link">dir.create</a></span><span class="op">(</span><span class="va">telltale_outdir</span>, showWarnings <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="finding-tales-loci-on-dna-sequences">Finding TALEs loci on DNA sequences<a class="anchor" aria-label="anchor" href="#finding-tales-loci-on-dna-sequences"></a>
</h2>
<p>We will run <code>telltale()</code> without correction for “gold” genomes, MAI1, BAI3 and PXO86. For “error-prone” genome BAI3-1-1 <code>telltale()</code> will be run with correction mode on.<br><code>telltale()</code> adds NTERM and CTERM to RVD sequences as markers of the termini, that is important for <a href="https://doi.org/10.3389/fpls.2015.00545" class="external-link">DisTal</a> to mach repeat codes with their cognate RVDs.</p>
<div class="section level3">
<h3 id="run-telltale-no-correction">Run telltale no correction<a class="anchor" aria-label="anchor" href="#run-telltale-no-correction"></a>
</h3>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/parallel/mclapply.html" class="external-link">mclapply</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">mai1_fa</span>, <span class="va">bai3_fa</span>, <span class="va">pxo86</span><span class="op">)</span>, <span class="kw">function</span><span class="op">(</span><span class="va">fastaFile</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">outdirname</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".fa"</span>, <span class="st">"_nocorrection"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">fastaFile</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">outputDir</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">telltale_outdir</span>, <span class="va">outdirname</span><span class="op">)</span></span>
<span>  <span class="fu"><a href="../reference/tellTale.html">tellTale</a></span><span class="op">(</span>subjectFile <span class="op">=</span> <span class="va">fastaFile</span>,</span>
<span>            outputDir <span class="op">=</span> <span class="va">outputDir</span>,</span>
<span>            appendExtremityCodes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>            rvdSep <span class="op">=</span> <span class="st">"-"</span>,</span>
<span>            talArrayCorrection <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span>
<span><span class="op">}</span>, mc.cores <span class="op">=</span> <span class="fl">3</span><span class="op">)</span></span>
<span><span class="co">#&gt;&gt; [[1]]</span></span>
<span><span class="co">#&gt;&gt; [1] "/home/cunnac/TEMP/test_tantale/telltale/MAI1_nocorrection"</span></span>
<span><span class="co">#&gt;&gt; </span></span>
<span><span class="co">#&gt;&gt; [[2]]</span></span>
<span><span class="co">#&gt;&gt; [1] "/home/cunnac/TEMP/test_tantale/telltale/BAI3_nocorrection"</span></span>
<span><span class="co">#&gt;&gt; </span></span>
<span><span class="co">#&gt;&gt; [[3]]</span></span>
<span><span class="co">#&gt;&gt; [1] "/home/cunnac/TEMP/test_tantale/telltale/PXO86_nocorrection"</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="run-telltale-correction">Run telltale correction<a class="anchor" aria-label="anchor" href="#run-telltale-correction"></a>
</h3>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/tellTale.html">tellTale</a></span><span class="op">(</span>subjectFile <span class="op">=</span> <span class="va">bai311_fa</span>,</span>
<span>         outputDir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">telltale_outdir</span>, <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">".fa"</span>, <span class="st">"_correction"</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="va">bai311_fa</span><span class="op">)</span><span class="op">)</span><span class="op">)</span>,</span>
<span>         TALE_CtermDNAHitMinScore <span class="op">=</span> <span class="fl">300</span>,</span>
<span>         appendExtremityCodes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>         rvdSep <span class="op">=</span> <span class="st">"-"</span>,</span>
<span>         talArrayCorrection <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span>
<span><span class="co">#&gt;&gt; Finding the closest reference amino acid sequences:</span></span>
<span><span class="co">#&gt;&gt; ================================================================================</span></span>
<span><span class="co">#&gt;&gt; </span></span>
<span><span class="co">#&gt;&gt; Time difference of 1.02 secs</span></span>
<span><span class="co">#&gt;&gt; Assessing frameshifts in nucleotide sequences:</span></span>
<span><span class="co">#&gt;&gt; ================================================================================</span></span>
<span><span class="co">#&gt;&gt; </span></span>
<span><span class="co">#&gt;&gt; Time difference of 433.6 secs</span></span>
<span><span class="co">#&gt;&gt; #****************************************</span></span>
<span><span class="co">#&gt;&gt; #**   tellTale analysis done     **</span></span>
<span><span class="co">#&gt;&gt; Current date:   Sun Sep 17 18:41:18 2023</span></span>
<span><span class="co">#&gt;&gt; #_________Provided I/O parameters __________</span></span>
<span><span class="co">#&gt;&gt; File of subject DNA sequences:  /tmp/RtmpijgdtX/file886ef5bb14bcc</span></span>
<span><span class="co">#&gt;&gt; TALE N-term CDS region detection HMM file:  /tmp/RtmpuGQoVx/temp_libpath846d21d45c7d2/tantale/extdata/hmmProfile/Xo_TALE_Nterm_CDS_profile.hmm</span></span>
<span><span class="co">#&gt;&gt; TALE repeat unit CDS detection HMM file:    /tmp/RtmpuGQoVx/temp_libpath846d21d45c7d2/tantale/extdata/hmmProfile/Xo_TALE_repeat_CDS_profile.hmm</span></span>
<span><span class="co">#&gt;&gt; TALE C-term CDS region detection HMM file:  /tmp/RtmpuGQoVx/temp_libpath846d21d45c7d2/tantale/extdata/hmmProfile/Xo_TALE_Cterm_CDS_profile.hmm</span></span>
<span><span class="co">#&gt;&gt; Output directory:   /home/cunnac/TEMP/test_tantale/telltale/BAI3-1-1_correction</span></span>
<span><span class="co">#&gt;&gt; #____________Other parameters________________</span></span>
<span><span class="co">#&gt;&gt; TALE_NtermDNAHitMinScore    :   300</span></span>
<span><span class="co">#&gt;&gt; repeatDNAHitMinScore    :   20</span></span>
<span><span class="co">#&gt;&gt; TALE_CtermDNAHitMinScore    :   300</span></span>
<span><span class="co">#&gt;&gt; minDomainHitsPerSubjSeq :   4</span></span>
<span><span class="co">#&gt;&gt; mergeHits   :   TRUE</span></span>
<span><span class="co">#&gt;&gt; minGapWidth :   35</span></span>
<span><span class="co">#&gt;&gt; extendedLength  :   300</span></span>
<span><span class="co">#&gt;&gt; talArrayCorrection  :   TRUE</span></span>
<span><span class="co">#&gt;&gt; refForTalArrayCorrection    :   /tmp/RtmpuGQoVx/temp_libpath846d21d45c7d2/tantale/extdata/decipher_ref_tales_aa.fa.gz</span></span>
<span><span class="co">#&gt;&gt; frameShiftCorrection    :   -11</span></span>
<span><span class="co">#&gt;&gt; #__________Summary measures of TALE search outcome__________</span></span>
<span><span class="co">#&gt;&gt; Number of analysed subject sequences :  2</span></span>
<span><span class="co">#&gt;&gt; Total number of TALE repeat DNA coding sequence motif hits found with the nhmmer approach:  155</span></span>
<span><span class="co">#&gt;&gt; Total number of subject seqs with TALE motif hits after low hit number filtering:   1</span></span>
<span><span class="co">#&gt;&gt; Total number of distinct regions (repeat arrays) with adjacent TALE motifs :    9</span></span>
<span><span class="co">#&gt;&gt; Total number of 'complete' arrays (with both N- and C-term flanking motifs):    8</span></span>
<span><span class="co">#&gt;&gt; Minimum array length (number of TALE domain hits):  1</span></span>
<span><span class="co">#&gt;&gt; Maximum array length:   28</span></span>
<span><span class="co">#&gt;&gt; Median array length:    20</span></span>
<span><span class="co">#&gt;&gt; Number of gaps of size below 500nt between TALE motifs arrays:  1.5</span></span>
<span><span class="co">#&gt;&gt; First quartile of size of gaps (below 500nt) between TALE motifs arrays:    108</span></span>
<span><span class="co">#&gt;&gt; Median size of gaps (below 500nt) between TALE motifs arrays:   108</span></span>
<span><span class="co">#&gt;&gt; Upper quartile of size of gaps (below 500nt) between TALE motifs arrays:    108</span></span>
<span><span class="co">#&gt;&gt; #__________Noteworthy AnnoTale issues__________</span></span>
<span><span class="co">#&gt;&gt; # </span></span>
<span><span class="co">#&gt;&gt; #*************************</span></span></code></pre></div>
</div>
<div class="section level3">
<h3 id="overview-of-telltale-results">Overview of telltale results<a class="anchor" aria-label="anchor" href="#overview-of-telltale-results"></a>
</h3>
<p>Besides RVD sequences and TALE ORFs, <code>telltale()</code> outputs other useful files such as the reports, from which we can get information about the structure of TALE sequences, the corrections that have been made, and so on.</p>
<p>First, we should take a look is “arrayReport.tsv”. We will take the files from 3 result folders, gather them, and see which information we can display from it.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># color set for ploting</span></span>
<span><span class="va">sunset10</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"#ba6b57"</span>, <span class="st">"#87556f"</span>, <span class="st">"#848ccf"</span>, <span class="st">"#93b5e1"</span>, <span class="st">"#6f4a8e"</span>, <span class="st">"#a2738c"</span>, <span class="st">"#ab93c9"</span>, <span class="st">"#588da8"</span>, <span class="st">"#726a95"</span>, <span class="st">"#bc658d"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">arrayReport_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">list.files</a></span><span class="op">(</span><span class="va">telltale_outdir</span>, <span class="st">"arrayReport.tsv"</span>, recursive <span class="op">=</span> <span class="cn">T</span>, full.names <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">arrayRp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/tibble.html" class="external-link">tibble</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">f</span> <span class="kw">in</span> <span class="va">arrayReport_files</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://readr.tidyverse.org/reference/read_delim.html" class="external-link">read_tsv</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="op">!</span><span class="fu"><a href="https://rdrr.io/r/base/any.html" class="external-link">any</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"predicted_ins_count"</span>, <span class="st">"predicted_dels_count"</span><span class="op">)</span> <span class="op"><a href="https://rdrr.io/pkg/BiocGenerics/man/match.html" class="external-link">%in%</a></span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">a</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>    <span class="co"># 2 columns "predicted_ins_count" and "predicted_dels_count" are not displayed in the report in case of no correction</span></span>
<span>    <span class="op">{</span><span class="va">a</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"predicted_ins_count"</span>, <span class="st">"predicted_dels_count"</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="cn">NA</span><span class="op">}</span> </span>
<span>  <span class="va">a</span><span class="op">$</span><span class="va">strain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\_.+"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">dirname</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">arrayRp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">arrayRp</span>, <span class="va">a</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="co">#&gt;&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">10</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">15</span></span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #BB0000;">chr</span> (6): arrayID, OriginalSubjectName, Strand, ArraySeq, SeqOfRVD, LongestOR...</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BB00;">dbl</span> (7): Start, End, NumberOfHits, N.terminusAAlength, C.terminusAAlength, L...</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #BBBB00;">lgl</span> (2): AllDomains, aberrantRepeat</span></span>
<span><span class="co">#&gt;&gt; </span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span><span class="co">#&gt;&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">9</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">17</span></span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #BB0000;">chr</span> (6): arrayID, OriginalSubjectName, Strand, ArraySeq, SeqOfRVD, LongestOR...</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BB00;">dbl</span> (9): Start, End, NumberOfHits, predicted_ins_count, predicted_dels_count...</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #BBBB00;">lgl</span> (2): AllDomains, aberrantRepeat</span></span>
<span><span class="co">#&gt;&gt; </span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span><span class="co">#&gt;&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">10</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">15</span></span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #BB0000;">chr</span> (6): arrayID, OriginalSubjectName, Strand, ArraySeq, SeqOfRVD, LongestOR...</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BB00;">dbl</span> (7): Start, End, NumberOfHits, N.terminusAAlength, C.terminusAAlength, L...</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #BBBB00;">lgl</span> (2): AllDomains, aberrantRepeat</span></span>
<span><span class="co">#&gt;&gt; </span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span><span class="co">#&gt;&gt; <span style="font-weight: bold;">Rows: </span><span style="color: #0000BB;">19</span> <span style="font-weight: bold;">Columns: </span><span style="color: #0000BB;">15</span></span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BBBB;">──</span> <span style="font-weight: bold;">Column specification</span> <span style="color: #00BBBB;">────────────────────────────────────────────────────────</span></span></span>
<span><span class="co">#&gt;&gt; <span style="font-weight: bold;">Delimiter:</span> "\t"</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #BB0000;">chr</span> (6): arrayID, OriginalSubjectName, Strand, ArraySeq, SeqOfRVD, LongestOR...</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BB00;">dbl</span> (7): Start, End, NumberOfHits, N.terminusAAlength, C.terminusAAlength, L...</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #BBBB00;">lgl</span> (2): AllDomains, aberrantRepeat</span></span>
<span><span class="co">#&gt;&gt; </span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BBBB;">ℹ</span> Use `spec()` to retrieve the full column specification for this data.</span></span>
<span><span class="co">#&gt;&gt; <span style="color: #00BBBB;">ℹ</span> Specify the column types or set `show_col_types = FALSE` to quiet this message.</span></span>
<span><span class="va">arrayRp</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://tibble.tidyverse.org/reference/as_tibble.html" class="external-link">as_tibble</a></span><span class="op">(</span><span class="va">arrayRp</span><span class="op">)</span></span></code></pre></div>
<p>This report takes result of HMMer for Tals DNA signatures detection, the results of the correction step and the AnnoTALE analyze step.<br>
What information does it contain:</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">arrayRp</span><span class="op">)</span></span>
<span><span class="co">#&gt;&gt;  [1] "arrayID"               "OriginalSubjectName"   "Start"                </span></span>
<span><span class="co">#&gt;&gt;  [4] "End"                   "Strand"                "NumberOfHits"         </span></span>
<span><span class="co">#&gt;&gt;  [7] "ArraySeq"              "AllDomains"            "SeqOfRVD"             </span></span>
<span><span class="co">#&gt;&gt; [10] "aberrantRepeat"        "N.terminusAAlength"    "C.terminusAAlength"   </span></span>
<span><span class="co">#&gt;&gt; [13] "LongestOrfLength"      "OrfCovOverArrayLength" "LongestORFSeq"        </span></span>
<span><span class="co">#&gt;&gt; [16] "predicted_ins_count"   "predicted_dels_count"  "strain"</span></span></code></pre></div>
<p>Some important points:</p>
<table class="table">
<colgroup>
<col width="26%">
<col width="73%">
</colgroup>
<thead><tr class="header">
<th>Column name</th>
<th>Content</th>
</tr></thead>
<tbody>
<tr class="odd">
<td><code>AllDomains</code></td>
<td>TRUE/FALSE, whether all the 3 domains of TALE arrays are detected</td>
</tr>
<tr class="even">
<td><code>aberrantRepeat</code></td>
<td>TRUE/FALSE, whether there is any aberrant repeat</td>
</tr>
<tr class="odd">
<td><code>predicted_ins_count</code></td>
<td>number of insertions corrected by <code>telltale()</code>
</td>
</tr>
<tr class="even">
<td><code>predicted_dels_count</code></td>
<td>number of deletions corrected by <code>telltale()</code>
</td>
</tr>
</tbody>
</table>
<div class="section level4">
<h4 id="count-alldomains-and-aberrantrepeat">Count AllDomains and aberrantRepeat<a class="anchor" aria-label="anchor" href="#count-alldomains-and-aberrantrepeat"></a>
</h4>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">arrayRp</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">strain</span>, fill <span class="op">=</span> <span class="va">AllDomains</span>, color <span class="op">=</span> <span class="va">aberrantRepeat</span><span class="op">)</span>, stat <span class="op">=</span> <span class="st">"count"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"grey"</span>, <span class="va">sunset10</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"black"</span>, <span class="st">"red"</span><span class="op">)</span>, na.value <span class="op">=</span> <span class="cn">NA</span>, labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"FALSE"</span>, <span class="st">"TRUE"</span>, <span class="st">"AnnoTALE_failed"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>color <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>override.aes <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">sunset10</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="1_tale_mining_files/figure-html/AllDomains%20and%20aberrantRepeat-1.png" width="576"></p>
</div>
<div class="section level4">
<h4 id="count-corrections">Count corrections<a class="anchor" aria-label="anchor" href="#count-corrections"></a>
</h4>
<p>total correction count per genome (in this case, there is only 1 genome that has been corrected)</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">arrayRp_correction</span> <span class="op">&lt;-</span> <span class="fu">reshape2</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/reshape2/man/melt.html" class="external-link">melt</a></span><span class="op">(</span><span class="va">arrayRp</span><span class="op">[</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"strain"</span>, <span class="st">"arrayID"</span>, <span class="st">"predicted_ins_count"</span>, <span class="st">"predicted_dels_count"</span><span class="op">)</span><span class="op">]</span>, value.name <span class="op">=</span> <span class="st">"count"</span>, variable.name <span class="op">=</span> <span class="st">"type"</span><span class="op">)</span></span>
<span><span class="co">#&gt;&gt; Using strain, arrayID as id variables</span></span>
<span><span class="va">arrayRp_correction</span> <span class="op"><a href="https://magrittr.tidyverse.org/reference/compound.html" class="external-link">%&lt;&gt;%</a></span> <span class="fu">dplyr</span><span class="fu">::</span><span class="fu"><a href="https://dplyr.tidyverse.org/reference/arrange.html" class="external-link">arrange</a></span><span class="op">(</span><span class="va">strain</span>, <span class="va">arrayID</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">arrayRp_correction</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">type</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span>, scales <span class="op">=</span> <span class="st">"fixed"</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">strain</span>, fill <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">count</span><span class="op">)</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_viridis.html" class="external-link">scale_fill_viridis_d</a></span><span class="op">(</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guides.html" class="external-link">guides</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/guide_legend.html" class="external-link">guide_legend</a></span><span class="op">(</span>title <span class="op">=</span> <span class="st">"correction count"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="1_tale_mining_files/figure-html/correction%20count%20per%20genome%202-1.png" width="576"></p>
<p>correction count per array per genome</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">arrayRp_correction</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">arrayID</span>, y <span class="op">=</span> <span class="va">count</span>, color <span class="op">=</span> <span class="va">type</span>, shape <span class="op">=</span> <span class="va">type</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">strain</span><span class="op">)</span>, nrow <span class="op">=</span> <span class="fl">2</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_shape_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">20</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_point.html" class="external-link">geom_point</a></span><span class="op">(</span>size <span class="op">=</span> <span class="fl">4</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sunset10</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>, <span class="st">"black"</span><span class="op">)</span><span class="op">)</span><span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>axis.text.x <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_text</a></span><span class="op">(</span>angle <span class="op">=</span> <span class="fl">90</span><span class="op">)</span>,</span>
<span>        legend.direction <span class="op">=</span> <span class="st">"horizontal"</span>, legend.position <span class="op">=</span> <span class="st">"bottom"</span><span class="op">)</span></span></code></pre></div>
<p><img src="1_tale_mining_files/figure-html/count%20per%20array%20per%20genome-1.png" width="576"></p>
<p>correction distribution per genome</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/stats/na.fail.html" class="external-link">na.omit</a></span><span class="op">(</span><span class="va">arrayRp_correction</span><span class="op">)</span>, <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">strain</span>, y <span class="op">=</span> <span class="va">count</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_dotplot.html" class="external-link">geom_dotplot</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>fill <span class="op">=</span> <span class="va">type</span>, color <span class="op">=</span> <span class="va">type</span><span class="op">)</span>, binaxis <span class="op">=</span> <span class="st">"y"</span>, stackdir <span class="op">=</span> <span class="st">"center"</span>, position <span class="op">=</span> <span class="st">"dodge"</span>, binwidth <span class="op">=</span> <span class="fl">1</span><span class="op">/</span><span class="fl">10</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sunset10</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>, <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_color_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">sunset10</span><span class="op">[</span><span class="fl">4</span><span class="op">]</span>, <span class="st">"black"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/theme.html" class="external-link">theme</a></span><span class="op">(</span>legend.direction <span class="op">=</span> <span class="st">"horizontal"</span>, legend.position <span class="op">=</span> <span class="st">"bottom"</span>,</span>
<span>        panel.grid.minor.y <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/element.html" class="external-link">element_blank</a></span><span class="op">(</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="1_tale_mining_files/figure-html/correction%20distribution-1.png" width="576"></p>
</div>
<div class="section level4">
<h4 id="count-domain-codon">Count domain codon<a class="anchor" aria-label="anchor" href="#count-domain-codon"></a>
</h4>
<p>Another report that we can reveal is “hitsReport.tsv”, which contains results of HMMer.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">hitsReport_files</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="va">telltale_outdir</span>, <span class="st">"hitsReport.tsv"</span>, recursive <span class="op">=</span> <span class="cn">T</span>, full.names <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">hitsReport</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">f</span> <span class="kw">in</span> <span class="va">hitsReport_files</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">f</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, stringsAsFactors <span class="op">=</span> <span class="cn">F</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>  <span class="va">a</span><span class="op">$</span><span class="va">strain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\_.+"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">dirname</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">hitsReport</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">hitsReport</span>, <span class="va">a</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">hitsReport</span><span class="op">$</span><span class="va">query_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">hitsReport</span><span class="op">$</span><span class="va">query_name</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"TALE_N-terminus_CDS_aligned_curated"</span>, <span class="st">"TALE_repeat_CDS_aligned_curated"</span>, <span class="st">"TALE_C-terminus_CDS_aligned_curated_long"</span><span class="op">)</span>, ordered <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p>information in the hitsReport</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/row_colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">hitsReport</span><span class="op">)</span></span>
<span><span class="co">#&gt;&gt;  [1] "arrayID"          "seqnames"         "start"            "end"             </span></span>
<span><span class="co">#&gt;&gt;  [5] "width"            "strand"           "nhmmerHitID"      "query_name"      </span></span>
<span><span class="co">#&gt;&gt;  [9] "hitID"            "seq"              "codon_count"      "frameshift_count"</span></span>
<span><span class="co">#&gt;&gt; [13] "strain"</span></span></code></pre></div>
<p>We could see the length distribution of each TALE protein domain (<code>query_name</code>) predicted by HMMer.</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">hitsReport</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">query_name</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"free"</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">codon_count</span><span class="op">)</span>, fill <span class="op">=</span> <span class="va">strain</span><span class="op">)</span>, stat <span class="op">=</span> <span class="st">"count"</span>, position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html" class="external-link">position_dodge</a></span><span class="op">(</span>preserve <span class="op">=</span> <span class="st">"single"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">sunset10</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_sqrt</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">15</span>, <span class="fl">100</span>, <span class="fl">200</span>, <span class="fl">350</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="1_tale_mining_files/figure-html/codon_count_annotale_3-1.png" width="960"></p>
<p>It seems that BAI3-1-1 has 1 TALE truncated at C-terminus caused by artifactual indels. But this is just the “raw” result of TALE motif detection.</p>
<p>Now we have a look at “domainsReport.tsv”, a similar report as “hitsReport.tsv” except that it contains the information of TALE domains predicted by AnnoTALE after correction, , so that we can make comparision and infer the impact of correction in the protein sequences.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">domainsReport_file</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.files.html" class="external-link">dir</a></span><span class="op">(</span><span class="va">telltale_outdir</span>, <span class="st">"domainsReport.tsv"</span>, recursive <span class="op">=</span> <span class="cn">T</span>, full.names <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span></span>
<span><span class="va">domainsReport</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">f</span> <span class="kw">in</span> <span class="va">domainsReport_file</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="va">a</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/utils/read.table.html" class="external-link">read.table</a></span><span class="op">(</span><span class="va">f</span>, sep <span class="op">=</span> <span class="st">"\t"</span>, stringsAsFactors <span class="op">=</span> <span class="cn">F</span>, header <span class="op">=</span> <span class="cn">T</span><span class="op">)</span></span>
<span>  <span class="va">a</span><span class="op">$</span><span class="va">strain</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span><span class="st">"\\_.+"</span>, <span class="st">""</span>, <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">basename</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/path.html" class="external-link">dirname</a></span><span class="op">(</span><span class="va">f</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span>  <span class="va">domainsReport</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/BiocGenerics/man/cbind.html" class="external-link">rbind</a></span><span class="op">(</span><span class="va">domainsReport</span>, <span class="va">a</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span></span>
<span><span class="va">domainsReport</span><span class="op">$</span><span class="va">query_name</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">factor</a></span><span class="op">(</span><span class="va">domainsReport</span><span class="op">$</span><span class="va">query_name</span>, levels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"N-terminus"</span>, <span class="st">"repeat"</span>, <span class="st">"C-terminus"</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">domainsReport</span><span class="op">$</span><span class="va">codon_count</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/factor.html" class="external-link">as.factor</a></span><span class="op">(</span><span class="va">domainsReport</span><span class="op">$</span><span class="va">codon_count</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/ggplot.html" class="external-link">ggplot</a></span><span class="op">(</span><span class="va">domainsReport</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/facet_wrap.html" class="external-link">facet_wrap</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/vars.html" class="external-link">vars</a></span><span class="op">(</span><span class="va">query_name</span><span class="op">)</span>, scales <span class="op">=</span> <span class="st">"free"</span>, nrow <span class="op">=</span> <span class="fl">1</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/geom_bar.html" class="external-link">geom_bar</a></span><span class="op">(</span><span class="fu"><a href="https://ggplot2.tidyverse.org/reference/aes.html" class="external-link">aes</a></span><span class="op">(</span>x <span class="op">=</span> <span class="va">codon_count</span>, fill <span class="op">=</span> <span class="va">strain</span><span class="op">)</span>, stat <span class="op">=</span> <span class="st">"count"</span>, position <span class="op">=</span> <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/position_dodge.html" class="external-link">position_dodge</a></span><span class="op">(</span>preserve <span class="op">=</span> <span class="st">"single"</span><span class="op">)</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_manual.html" class="external-link">scale_fill_manual</a></span><span class="op">(</span>values <span class="op">=</span> <span class="va">sunset10</span><span class="op">)</span> <span class="op">+</span></span>
<span>  <span class="fu"><a href="https://ggplot2.tidyverse.org/reference/scale_continuous.html" class="external-link">scale_y_sqrt</a></span><span class="op">(</span>breaks <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">1</span>, <span class="fl">5</span>, <span class="fl">10</span>, <span class="fl">15</span>, <span class="fl">100</span>, <span class="fl">200</span>, <span class="fl">350</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p><img src="1_tale_mining_files/figure-html/codon_count_hmm-1.png" width="960"></p>
<p>Firstly, the domain detection of AnnoTALE is different from HMMer at some amino acids.<br>
Secondly, the truncated C-terminus of BAI3-1-1 (84 aa) has been corrected and has the same length as its counterparts.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/save.html" class="external-link">save.image</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/file.path.html" class="external-link">file.path</a></span><span class="op">(</span><span class="va">outdir</span>, <span class="st">"mining.RData"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Bao Tram Vi, Sebastien Cunnac.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
