<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="tantale"><title>3. Multiple alignment of TALE sequences with tantale • tantale</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="3. Multiple alignment of TALE sequences with tantale"><meta property="og:description" content="tantale"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">tantale</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.9550</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/1_tale_mining.html">1. Mining TALE sequences in genomes with tantale</a>
    <a class="dropdown-item" href="../articles/2_tale_classification.html">2. Classifying TALE sequences from genomes with tantale</a>
    <a class="dropdown-item" href="../articles/3_tale_msa.html">3. Multiple alignment of TALE sequences with tantale</a>
    <a class="dropdown-item" href="../articles/4_tale_target_prediction.html">4. TALE targets predictions with tantale</a>
    <a class="dropdown-item" href="../articles/p1_distalr.html">A comparison between `runDistal` and `distalr` implementations</a>
    <a class="dropdown-item" href="../articles/p2_multiple_alignments.html">TALEs domains sequences alignment with MAFFT</a>
    <a class="dropdown-item" href="../articles/p3_tantale_objects.html">An overview of the various objects related to TALEs in tantale</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/scunnac/tantale/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-article">



<script src="3_tale_msa_files/htmlwidgets-1.6.2/htmlwidgets.js"></script>
<link href="3_tale_msa_files/datatables-css-0.0.0/datatables-crosstalk.css" rel="stylesheet" />
<script src="3_tale_msa_files/datatables-binding-0.29/datatables.js"></script>
<link href="3_tale_msa_files/dt-core-1.13.4/css/jquery.dataTables.min.css" rel="stylesheet" />
<link href="3_tale_msa_files/dt-core-1.13.4/css/jquery.dataTables.extra.css" rel="stylesheet" />
<script src="3_tale_msa_files/dt-core-1.13.4/js/jquery.dataTables.min.js"></script>
<link href="3_tale_msa_files/crosstalk-1.2.0/css/crosstalk.min.css" rel="stylesheet" />
<script src="3_tale_msa_files/crosstalk-1.2.0/js/crosstalk.min.js"></script>

<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>3. Multiple alignment of TALE sequences with tantale</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/scunnac/tantale/blob/HEAD/vignettes/3_tale_msa.Rmd" class="external-link"><code>vignettes/3_tale_msa.Rmd</code></a></small>
      <div class="d-none name"><code>3_tale_msa.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" tabindex="-1"></a><span class="fu">library</span>(tantale)</span>
<span id="cb1-2"><a href="#cb1-2" tabindex="-1"></a><span class="fu">library</span>(ggplot2)</span>
<span id="cb1-3"><a href="#cb1-3" tabindex="-1"></a><span class="fu">library</span>(tidyverse)</span>
<span id="cb1-4"><a href="#cb1-4" tabindex="-1"></a><span class="fu">library</span>(Biostrings)</span>
<span id="cb1-5"><a href="#cb1-5" tabindex="-1"></a><span class="fu">library</span>(parallel)</span>
<span id="cb1-6"><a href="#cb1-6" tabindex="-1"></a><span class="fu">library</span>(magrittr)</span>
<span id="cb1-7"><a href="#cb1-7" tabindex="-1"></a><span class="fu">library</span>(DT)</span>
<span id="cb1-8"><a href="#cb1-8" tabindex="-1"></a><span class="fu">library</span>(gplots)</span>
<span id="cb1-9"><a href="#cb1-9" tabindex="-1"></a><span class="fu">library</span>(ape)</span>
<span id="cb1-10"><a href="#cb1-10" tabindex="-1"></a><span class="fu">library</span>(ggtree)</span>
<span id="cb1-11"><a href="#cb1-11" tabindex="-1"></a><span class="fu">library</span>(corrr)</span>
<span id="cb1-12"><a href="#cb1-12" tabindex="-1"></a><span class="fu">library</span>(ggcorrplot)</span></code></pre></div>
<p>When investigating large sets of TALEs with groups composed of
several members, it is often quite informative to observe a multiple
alignment of the ‘parts’ of the TALEs belonging to groups of
interest.</p>
<p>The TALE Class Assignment tool of AnnoTALE performs this task but one
drawback is that it is unable to insert gaps in the alignements. New
TALEs often evolve by DNA recombination creating large deletion and
insertions in the Central Repeat Domain of the protein. It is therefore
very powerful to be able to detect inDels because they may be the sign
of common descent (see also the article on the <a
href="https://doi.org/10.3389/fpls.2015.00545">QueTal suite</a>).</p>
<p>Under the hood, the <code>buildRepeatMsa</code> function leverages
the capacity of <a
href="https://mafft.cbrc.jp/alignment/software/">MAFFT</a> to align
arbitrary characters in hexadecimal code to generate TALE parts
sequences multiple alignments as illustrated here.</p>
<div class="sourceCode" id="cb2"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" tabindex="-1"></a>outdir <span class="ot">&lt;-</span> fs<span class="sc">::</span><span class="fu">dir_create</span>(<span class="st">&quot;~/TEMP/test_tantale&quot;</span>) <span class="co"># tempdir(check = TRUE)</span></span>
<span id="cb2-2"><a href="#cb2-2" tabindex="-1"></a><span class="fu">load</span>(<span class="fu">file.path</span>(outdir, <span class="st">&quot;mining.RData&quot;</span>))</span></code></pre></div>
<div id="tale-domains-multiple-alignments-with-similarity"
class="section level1">
<h1>TALE domains multiple alignments with similarity</h1>
<p>If interested in performing TALEs multiple alignments (distal domains
or RVDs) it makes sense to align TALEs that are related and that thus
belong to the same group. This is what we are doing here but keep in
mind that the <code>buildRepeatMsa()</code> function can take any
arbitrary set of repeatID sequences or RVD sequences.</p>
<p>But right now what we need is a mapping between distal parts IDs and
the cognate RVDs. This can be done in a number of ways but let’s use the
one below that rely on the `run``.</p>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" tabindex="-1"></a></span>
<span id="cb3-2"><a href="#cb3-2" tabindex="-1"></a>rvdSeqs_fasta <span class="ot">&lt;-</span> <span class="fu">list.files</span>(telltale_outdir, <span class="st">&quot;rvdSequences.fas&quot;</span>, <span class="at">recursive =</span> T, <span class="at">full.names =</span> T)</span>
<span id="cb3-3"><a href="#cb3-3" tabindex="-1"></a>telltale_rvd <span class="ot">&lt;-</span> <span class="fu">file.path</span>(distal_telltale, <span class="st">&quot;rvd.fasta&quot;</span>)</span>
<span id="cb3-4"><a href="#cb3-4" tabindex="-1"></a><span class="fu">unlink</span>(telltale_rvd)</span>
<span id="cb3-5"><a href="#cb3-5" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" tabindex="-1"></a><span class="co"># rename of RVD sequences to distinguish sequences between different strains</span></span>
<span id="cb3-7"><a href="#cb3-7" tabindex="-1"></a><span class="cf">for</span> (input <span class="cf">in</span> rvdSeqs_fasta) {</span>
<span id="cb3-8"><a href="#cb3-8" tabindex="-1"></a>  id <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;</span><span class="sc">\\</span><span class="st">_.+&quot;</span>, <span class="st">&quot;&quot;</span>, <span class="fu">basename</span>(<span class="fu">dirname</span>(input)))</span>
<span id="cb3-9"><a href="#cb3-9" tabindex="-1"></a>  seqs <span class="ot">&lt;-</span> <span class="fu">readBStringSet</span>(input, <span class="at">seek.first.rec =</span> T)</span>
<span id="cb3-10"><a href="#cb3-10" tabindex="-1"></a>  <span class="fu">names</span>(seqs) <span class="ot">&lt;-</span> <span class="fu">paste0</span>(id, <span class="st">&quot;_&quot;</span>, <span class="fu">names</span>(seqs))</span>
<span id="cb3-11"><a href="#cb3-11" tabindex="-1"></a>  <span class="fu">names</span>(seqs) <span class="ot">&lt;-</span> stringr<span class="sc">::</span><span class="fu">str_trim</span>(<span class="fu">names</span>(seqs))</span>
<span id="cb3-12"><a href="#cb3-12" tabindex="-1"></a>  seqs <span class="ot">&lt;-</span> seqs[<span class="fu">width</span>(seqs) <span class="sc">!=</span> <span class="dv">0</span>]</span>
<span id="cb3-13"><a href="#cb3-13" tabindex="-1"></a>  <span class="fu">writeXStringSet</span>(seqs, telltale_rvd, <span class="at">append =</span> T)</span>
<span id="cb3-14"><a href="#cb3-14" tabindex="-1"></a>}</span>
<span id="cb3-15"><a href="#cb3-15" tabindex="-1"></a></span>
<span id="cb3-16"><a href="#cb3-16" tabindex="-1"></a>rvdVectors <span class="ot">&lt;-</span> <span class="fu">toListOfSplitedStr</span>(telltale_rvd, <span class="at">sep =</span> <span class="st">&quot;-&quot;</span>)</span>
<span id="cb3-17"><a href="#cb3-17" tabindex="-1"></a>repeatVectors <span class="ot">&lt;-</span> <span class="fu">toListOfSplitedStr</span>(distalr_deci_output<span class="sc">$</span>coded.repeats.str, <span class="at">sep =</span> <span class="st">&quot; &quot;</span>)</span>
<span id="cb3-18"><a href="#cb3-18" tabindex="-1"></a></span>
<span id="cb3-19"><a href="#cb3-19" tabindex="-1"></a><span class="co"># map repeatID to RVD</span></span>
<span id="cb3-20"><a href="#cb3-20" tabindex="-1"></a>rep2rvd <span class="ot">&lt;-</span> <span class="fu">getRepeat2RvdMapping</span>(<span class="at">talesRepeatVectors =</span> repeatVectors, <span class="at">talesRvdVectors =</span> rvdVectors)</span>
<span id="cb3-21"><a href="#cb3-21" tabindex="-1"></a><span class="fu">datatable</span>(rep2rvd, <span class="at">options =</span> <span class="fu">list</span>(<span class="at">pageLength =</span> <span class="dv">5</span>))</span></code></pre></div>
<div class="datatables html-widget html-fill-item-overflow-hidden html-fill-item" id="htmlwidget-29d29edb66c10ed97587" style="width:100%;height:auto;"></div>
<script type="application/json" data-for="htmlwidget-29d29edb66c10ed97587">{"x":{"filter":"none","vertical":false,"data":[["1","2","3","4","5","6","7","8","9","10","11","12","13","14","15","16","17","18","19","20","21","22","23","24","25","26","27","28","29","30","31","32","33","34","35","36","37","38","39","40","41","42","43","44","45","46","47","48","49","50","51","52","53","54","55","56","57","58","59","60","61","62","63","64","65","66","67","68","69","70","71","72","73","74","75","76","77","78","79","80","81","82","83","84","85","86","87","88","89","90","91","92","93","94","95","96","97","98","99","100","101","102","103","104","105","106","107","108","109","110","111","112","113","114","115","116","117","118","119","120","121","122","123","124","125","126","127","128","129","130","131","132","133","134","135","136","137","138","139","140","141","142","143","144","145","146","147","148","149","150","151","152","153","154","155","156","157","158","159","160","161","162","163","164","165","166","167","168","169","170","171","172","173","174","175","176","177","178","179","180","181","182","183","184","185","186","187","188","189","190","191","192","193","194","195","196","197","198","199","200","201","202","203","204","205","206","207","208","209","210","211","212","213","214","215","216","217","218","219","220","221","222","223","224","225","226","227","228","229","230","231","232","233","234","235","236","237","238","239","240","241","242","243","244","245","246","247","248","249","250","251"],["1","10","100","101","102","103","104","105","106","107","108","109","11","110","111","112","113","114","115","116","117","118","119","12","120","121","122","123","124","125","126","127","128","129","13","130","131","132","133","134","135","136","137","138","139","14","140","141","142","143","144","145","146","147","148","149","15","150","151","152","153","154","155","156","157","158","159","16","160","161","162","163","164","165","166","167","168","169","17","170","171","172","173","174","175","176","177","178","179","18","180","181","182","183","184","185","186","187","188","189","19","190","191","192","193","194","195","196","197","198","199","2","20","200","201","202","203","204","205","206","207","208","209","21","210","211","212","213","214","215","216","217","218","219","22","220","221","222","223","224","225","226","227","228","229","23","230","231","232","233","234","235","236","237","238","239","24","240","241","242","243","244","245","246","247","248","249","25","250","251","26","27","28","29","3","30","31","32","33","34","35","36","37","38","39","4","40","41","42","43","44","45","46","47","48","49","5","50","51","52","53","54","55","56","57","58","59","6","60","61","62","63","64","65","66","67","68","69","7","70","71","72","73","74","75","76","77","78","79","8","80","81","82","83","84","85","86","87","88","89","9","90","91","92","93","94","95","96","97","98","99"],["N*","HD","HD","HD","HD","HD","HG","HG","HG","HG","HG","H*","HD","H*","H*","HG","NG","ng","NG","NG","N*","N*","N*","NS","N*","n*","NG","NI","NI","NI","NI","NI","NI","NI","NG","NI","NI","NI","NN","NN","NN","NN","ni","NS","NS","NI","NS","SS","NN","NS","NN","NG","HD","HD","NN","HD","N*","NG","NG","NG","N*","NI","NK","NN","NN","NN","NN","HD","NG","NG","NG","N*","NI","NI","NS","HD","HD","N*","HD","N*","NS","N*","HD","NI","NS","NN","HD","N*","HD","HD","NN","NTERM","NTERM","NTERM","NTERM","NTERM","NTERM","NTERM","NTERM","NTERM","HG","NTERM","NTERM","NTERM","NTERM","NTERM","NTERM","NTERM","NTERM","NTERM","NTERM","NN","ND","NTERM","NTERM","NTERM","NTERM","NTERM","NTERM","NTERM","NTERM","NTERM","NTERM","NG","NTERM","NTERM","NTERM","NTERM","NTERM","CTERM","CTERM","CTERM","CTERM","CTERM","N*","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","NG","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","NI","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","CTERM","NI","CTERM","XXXXX","HD","NG","NI","NI","NI","HD","HD","NG","NI","NN","NN","NN","HD","HD","HD","HD","HD","HD","HD","HD","HD","HG","HG","HI","ND","NG","NG","NG","NG","NG","N*","N*","NI","NI","NI","NI","NI","NI","NI","NN","NN","NN","NN","NN","NN","NS","NS","NS","NV","S*","HD","HD","NN","HD","NI","NN","NN","ND","NN","NI","NK","NN","NN","NN","NN","NN","NN","NN","NV","HD","NN","HD","HD","HD","HD","HD","HD","HD","HD","HD","HD"]],"container":"<table class=\"display\">\n  <thead>\n    <tr>\n      <th> <\/th>\n      <th>repeatID<\/th>\n      <th>RVD<\/th>\n    <\/tr>\n  <\/thead>\n<\/table>","options":{"pageLength":5,"columnDefs":[{"orderable":false,"targets":0}],"order":[],"autoWidth":false,"orderClasses":false,"lengthMenu":[5,10,25,50,100]},"selection":{"mode":"multiple","selected":null,"target":"row","selectable":null}},"evals":[],"jsHooks":[]}</script>
<div id="performing-msa-using-repeat-strings" class="section level2">
<h2>Performing MSA using repeat strings</h2>
<p>If possible it is more appropriate to align repeat/domains rather
than RVD. For mafft (the underlying aligner), in our hands, the combo of
gap opening score = 0 (<code>--op 0</code>) and gap extending score = 5
(<code>--ep 5</code>) (the current default values) yields better
alignments.</p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" tabindex="-1"></a>repeatVectors <span class="ot">&lt;-</span> distalr_deci_output<span class="sc">$</span>coded.repeats.str</span>
<span id="cb4-2"><a href="#cb4-2" tabindex="-1"></a>repeatSim <span class="ot">&lt;-</span> distalr_deci_output<span class="sc">$</span>repeat.similarity <span class="co">#distal_output$repeat.similarity</span></span>
<span id="cb4-3"><a href="#cb4-3" tabindex="-1"></a></span>
<span id="cb4-4"><a href="#cb4-4" tabindex="-1"></a></span>
<span id="cb4-5"><a href="#cb4-5" tabindex="-1"></a><span class="fu">lapply</span>(<span class="fu">sort</span>(<span class="fu">unique</span>(taleGroups<span class="sc">$</span>group)), <span class="cf">function</span>(G) {repeatVectors[taleGroups<span class="sc">$</span>name[taleGroups<span class="sc">$</span>group <span class="sc">==</span> G]]})</span>
<span id="cb4-6"><a href="#cb4-6" tabindex="-1"></a></span>
<span id="cb4-7"><a href="#cb4-7" tabindex="-1"></a>repeatMsaByGroup_withSim <span class="ot">&lt;-</span> <span class="fu">lapply</span>(<span class="fu">sort</span>(<span class="fu">unique</span>(taleGroups<span class="sc">$</span>group)), <span class="cf">function</span>(G) {</span>
<span id="cb4-8"><a href="#cb4-8" tabindex="-1"></a>  <span class="fu">try</span>({<span class="fu">buildRepeatMsa</span>(<span class="at">inputSeqs =</span> repeatVectors[taleGroups<span class="sc">$</span>name[taleGroups<span class="sc">$</span>group <span class="sc">==</span> G]],</span>
<span id="cb4-9"><a href="#cb4-9" tabindex="-1"></a>                      <span class="at">sep =</span> <span class="st">&quot; &quot;</span>,</span>
<span id="cb4-10"><a href="#cb4-10" tabindex="-1"></a>                      <span class="at">distalRepeatSims =</span> repeatSim,</span>
<span id="cb4-11"><a href="#cb4-11" tabindex="-1"></a>                      <span class="at">mafftOpts =</span> <span class="st">&quot;--globalpair --weighti 1 --maxiterate 1000 --quiet --reorder --op 0 --ep 5 --thread 1&quot;</span>)</span>
<span id="cb4-12"><a href="#cb4-12" tabindex="-1"></a>  })</span>
<span id="cb4-13"><a href="#cb4-13" tabindex="-1"></a>}</span>
<span id="cb4-14"><a href="#cb4-14" tabindex="-1"></a>)</span></code></pre></div>
</div>
<div
id="plotting-msa-with-rvds-as-text-and-coloring-based-on-repeat-similarity"
class="section level2">
<h2>Plotting MSA with RVDs as text and coloring based on repeat
similarity</h2>
<p>For this type of plot, we need to map repeatID to RVD with
<code>getRepeat2RvdMapping()</code>, then convert repeat MSA to RVD MSA
by <code>convertRepeat2RvdAlign</code>. We should use the converting
function rather than using <code>buildRepeatMsa()</code> to creat MSA of
RVD because there will be disagreement between them.</p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" tabindex="-1"></a><span class="co"># convert repeat MSA to RVD MSA</span></span>
<span id="cb5-2"><a href="#cb5-2" tabindex="-1"></a>repeatMSA <span class="ot">&lt;-</span> repeatMsaByGroup_withSim[[<span class="dv">2</span>]]</span>
<span id="cb5-3"><a href="#cb5-3" tabindex="-1"></a>rvdMSA <span class="ot">&lt;-</span>  <span class="fu">convertRepeat2RvdAlign</span>(<span class="at">repeatAlign =</span> repeatMSA, <span class="at">repeat2RvdMapping =</span> rep2rvd)</span>
<span id="cb5-4"><a href="#cb5-4" tabindex="-1"></a></span>
<span id="cb5-5"><a href="#cb5-5" tabindex="-1"></a><span class="co"># plot MSA</span></span>
<span id="cb5-6"><a href="#cb5-6" tabindex="-1"></a>tantale<span class="sc">::</span><span class="fu">heatmap_msa</span>(<span class="at">talsim =</span> distalr_deci_output<span class="sc">$</span>tal.similarity,</span>
<span id="cb5-7"><a href="#cb5-7" tabindex="-1"></a>                     <span class="at">repeatAlign =</span> repeatMSA,</span>
<span id="cb5-8"><a href="#cb5-8" tabindex="-1"></a>                     <span class="at">repeatSim =</span> distalr_deci_output<span class="sc">$</span>repeat.similarity,</span>
<span id="cb5-9"><a href="#cb5-9" tabindex="-1"></a>                     <span class="at">rvdAlign =</span> rvdMSA,</span>
<span id="cb5-10"><a href="#cb5-10" tabindex="-1"></a>                     <span class="at">plot.type =</span> <span class="st">&quot;repeat.similarity&quot;</span>,</span>
<span id="cb5-11"><a href="#cb5-11" tabindex="-1"></a>                     <span class="at">main =</span> <span class="st">&quot;Repeat MSA with similarity - Group 2&quot;</span>)</span></code></pre></div>
<p><img src="3_tale_msa_files/figure-html/repeat.similarity_plot-1.png" width="1152" /></p>
</div>
<div
id="plotting-msa-with-repeat-ids-as-text-and-coloring-based-on-repeat-clustering"
class="section level2">
<h2>Plotting MSA with repeat Ids as text and coloring based on repeat
clustering</h2>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb6-1"><a href="#cb6-1" tabindex="-1"></a>tantale<span class="sc">::</span><span class="fu">heatmap_msa</span>(<span class="at">talsim =</span> distalr_deci_output<span class="sc">$</span>tal.similarity,</span>
<span id="cb6-2"><a href="#cb6-2" tabindex="-1"></a>                     <span class="at">repeatAlign =</span> repeatMSA,</span>
<span id="cb6-3"><a href="#cb6-3" tabindex="-1"></a>                     <span class="at">repeatSim =</span> distalr_deci_output<span class="sc">$</span>repeat.similarity,</span>
<span id="cb6-4"><a href="#cb6-4" tabindex="-1"></a>                     <span class="at">plot.type =</span> <span class="st">&quot;repeat.clusters&quot;</span>,</span>
<span id="cb6-5"><a href="#cb6-5" tabindex="-1"></a>                     <span class="at">main =</span> <span class="st">&quot;Repeat Cluster MSA - Group 2&quot;</span>)</span></code></pre></div>
<p><img src="3_tale_msa_files/figure-html/repeat.clusters_plot-1.png" width="1152" /></p>
</div>
<div
id="plotting-msa-with-rvd-as-text-and-coloring-based-on-repeat-clustering"
class="section level2">
<h2>Plotting MSA with RVD as text and coloring based on repeat
clustering</h2>
<p>For this plot, we also use the RVD MSA converted from repeat MSA.</p>
<div class="sourceCode" id="cb7"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb7-1"><a href="#cb7-1" tabindex="-1"></a>tantale<span class="sc">::</span><span class="fu">heatmap_msa</span>(<span class="at">talsim =</span> distalr_deci_output<span class="sc">$</span>tal.similarity,</span>
<span id="cb7-2"><a href="#cb7-2" tabindex="-1"></a>                     <span class="at">repeatAlign =</span> repeatMSA,</span>
<span id="cb7-3"><a href="#cb7-3" tabindex="-1"></a>                     <span class="at">repeatSim =</span> distalr_deci_output<span class="sc">$</span>repeat.similarity,</span>
<span id="cb7-4"><a href="#cb7-4" tabindex="-1"></a>                     <span class="at">rvdAlign =</span> rvdMSA,</span>
<span id="cb7-5"><a href="#cb7-5" tabindex="-1"></a>                     <span class="at">plot.type =</span> <span class="st">&quot;repeat.clusters.with.rvd&quot;</span>,</span>
<span id="cb7-6"><a href="#cb7-6" tabindex="-1"></a>                     <span class="at">main =</span> <span class="st">&quot;Repeat Cluster MSA with RVD - Group 2&quot;</span>,</span>
<span id="cb7-7"><a href="#cb7-7" tabindex="-1"></a>                     <span class="at">consensusSeq =</span> <span class="cn">TRUE</span>)</span></code></pre></div>
<p><img src="3_tale_msa_files/figure-html/repeat.clusters.with.rvd_plot-1.png" width="1152" /></p>
</div>
</div>
<div id="generate-an-overview-of-talomes" class="section level1">
<h1>Generate an overview of talomes</h1>
<p>With the TALE classification result and RVD sequences, we can
represent RVD sequence variants of all the strains for each TALE group
as an example below.</p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb8-1"><a href="#cb8-1" tabindex="-1"></a>tale_annotation <span class="ot">&lt;-</span> <span class="fu">merge</span>(taleGroups, </span>
<span id="cb8-2"><a href="#cb8-2" tabindex="-1"></a>                         <span class="fu">data.frame</span>(<span class="st">&quot;name&quot;</span> <span class="ot">=</span> <span class="fu">names</span>(rvdVectors),</span>
<span id="cb8-3"><a href="#cb8-3" tabindex="-1"></a>                                    <span class="st">&quot;rvdseq&quot;</span> <span class="ot">=</span> <span class="fu">sapply</span>(rvdVectors,</span>
<span id="cb8-4"><a href="#cb8-4" tabindex="-1"></a>                                                      <span class="cf">function</span>(i) <span class="fu">paste</span>(i, <span class="at">collapse =</span> <span class="st">&quot;-&quot;</span>)</span>
<span id="cb8-5"><a href="#cb8-5" tabindex="-1"></a>                                                      ), <span class="at">stringsAsFactors =</span> <span class="cn">FALSE</span></span>
<span id="cb8-6"><a href="#cb8-6" tabindex="-1"></a>                                    )</span>
<span id="cb8-7"><a href="#cb8-7" tabindex="-1"></a>                         )</span>
<span id="cb8-8"><a href="#cb8-8" tabindex="-1"></a>tale_annotation<span class="sc">$</span>strain <span class="ot">&lt;-</span> <span class="fu">gsub</span>(<span class="st">&quot;</span><span class="sc">\\</span><span class="st">_.+&quot;</span>, <span class="st">&quot;&quot;</span>, tale_annotation<span class="sc">$</span>name)</span>
<span id="cb8-9"><a href="#cb8-9" tabindex="-1"></a></span>
<span id="cb8-10"><a href="#cb8-10" tabindex="-1"></a><span class="fu">heatmap_talomes</span>(tale_annotation,</span>
<span id="cb8-11"><a href="#cb8-11" tabindex="-1"></a>                <span class="at">col =</span> <span class="st">&quot;group&quot;</span>, <span class="at">row =</span> <span class="st">&quot;strain&quot;</span>,</span>
<span id="cb8-12"><a href="#cb8-12" tabindex="-1"></a>                <span class="at">value =</span> <span class="st">&quot;rvdseq&quot;</span>,</span>
<span id="cb8-13"><a href="#cb8-13" tabindex="-1"></a>                <span class="at">mapcol =</span> sunset10)</span></code></pre></div>
<p><img src="3_tale_msa_files/figure-html/talomes_overview-1.png" width="960" /></p>
<div class="sourceCode" id="cb9"><pre class="sourceCode r"><code class="sourceCode r"><span id="cb9-1"><a href="#cb9-1" tabindex="-1"></a><span class="fu">save.image</span>(<span class="fu">file.path</span>(outdir, <span class="st">&quot;mining.RData&quot;</span>))</span></code></pre></div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>



    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Bao Tram Vi, Sebastien Cunnac.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>
