<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="The function creates a graphical presentation from a tale annotation table. The output is like a heatmap that presents rvd sequence variants in Tal groups as column and respective strains as rows (or vice versa). It is different from a typical heatmap that it can display more than one value in a cell; for example, if one strain has 2 rvd sequence variants belong to 1 group, it will be displayed by 2 colors in 1 cell."><title>Heatmap plotting of rvd sequence variants — heatmap_talomes • tantale</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Heatmap plotting of rvd sequence variants — heatmap_talomes"><meta property="og:description" content="The function creates a graphical presentation from a tale annotation table. The output is like a heatmap that presents rvd sequence variants in Tal groups as column and respective strains as rows (or vice versa). It is different from a typical heatmap that it can display more than one value in a cell; for example, if one strain has 2 rvd sequence variants belong to 1 group, it will be displayed by 2 colors in 1 cell."><meta name="robots" content="noindex"><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">tantale</a>

    <small class="nav-text text-danger me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="In-development version">0.1.9550</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/1_tale_mining.html">1. Mining TALE sequences in genomes with tantale</a>
    <a class="dropdown-item" href="../articles/2_tale_classification.html">2. Classifying TALE sequences from genomes with tantale</a>
    <a class="dropdown-item" href="../articles/3_tale_msa.html">3. Multiple alignment of TALE sequences with tantale</a>
    <a class="dropdown-item" href="../articles/4_tale_target_prediction.html">4. TALE targets predictions with tantale</a>
    <a class="dropdown-item" href="../articles/p1_distalr.html">A comparison between `runDistal` and `distalr` implementations</a>
    <a class="dropdown-item" href="../articles/p2_multiple_alignments.html">TALEs domains sequences alignment with MAFFT</a>
    <a class="dropdown-item" href="../articles/p3_tantale_objects.html">An overview of the various objects related to TALEs in tantale</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/scunnac/tantale/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Heatmap plotting of rvd sequence variants</h1>
      <small class="dont-index">Source: <a href="https://github.com/scunnac/tantale/blob/HEAD/R/classification.R" class="external-link"><code>R/classification.R</code></a></small>
      <div class="d-none name"><code>heatmap_talomes.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p>The function creates a graphical presentation from a tale annotation table. The output is like a heatmap that presents rvd sequence variants in Tal groups as column and respective strains as rows (or vice versa). It is different from a typical heatmap that it can display more than one value in a cell; for example, if one strain has 2 rvd sequence variants belong to 1 group, it will be displayed by 2 colors in 1 cell.</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">heatmap_talomes</span><span class="op">(</span></span>
<span>  <span class="va">tale_annotation</span>,</span>
<span>  <span class="va">col</span>,</span>
<span>  <span class="va">row</span>,</span>
<span>  <span class="va">value</span>,</span>
<span>  truncTaleLab <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  extraCol <span class="op">=</span> <span class="cn">NULL</span>,</span>
<span>  x.lab <span class="op">=</span> <span class="st">"TALE Group"</span>,</span>
<span>  y.lab <span class="op">=</span> <span class="st">"Strain"</span>,</span>
<span>  title <span class="op">=</span> <span class="st">"RVD sequences variants"</span>,</span>
<span>  plot.type <span class="op">=</span> <span class="st">"all"</span>,</span>
<span>  mapcol <span class="op">=</span> <span class="fu">viridis</span><span class="fu">::</span><span class="fu"><a href="https://sjmgarnier.github.io/viridisLite/reference/viridis.html" class="external-link">viridis</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span>,</span>
<span>  mar.side <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">5</span>, <span class="fl">5</span>, <span class="fl">3</span>, <span class="fl">3</span><span class="op">)</span>,</span>
<span>  sepwid <span class="op">=</span> <span class="fl">5</span>,</span>
<span>  sepcol <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>  inner_sepcol <span class="op">=</span> <span class="st">"white"</span>,</span>
<span>  save.path <span class="op">=</span> <span class="cn">NULL</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>tale_annotation</dt>
<dd><p>a data frame containing at least 3 columns for Tal groups, strain names, and rvd seqs, and 1 row is 1 Tal.</p></dd>


<dt>col</dt>
<dd><p>"character", column name of <code>tale_annotation</code> to be displayed as rows in the heatmap (e.g. tal groups).</p></dd>


<dt>row</dt>
<dd><p>"character", column name of <code>tale_annotation</code> to be displayed as columns in the heatmap (e.g. strain names).</p></dd>


<dt>value</dt>
<dd><p>"character", column name for rvdseqs in the <code>tale_annotation</code></p></dd>


<dt>truncTaleLab</dt>
<dd><p>(optional, default = NULL) "character", column name of <code>tale_annotation</code> labeling the truncTales by TRUE/FALSE value. The truncTales are labeled by "T" in the heatmap cells, but if this argument is called.</p></dd>


<dt>extraCol</dt>
<dd><p>(optional, default = NULL) "character", column name of <code>tale_annotation</code> containing other information (e.g. origin). It will be presented in a side bar on the right of the heatmap.</p></dd>


<dt>x.lab, y.lab, title</dt>
<dd><p>character for x axix, y axis names and title</p></dd>


<dt>mapcol</dt>
<dd><p>character vector of colors for the cells.</p></dd>


<dt>mar.side</dt>
<dd><p>margin of the heatmap for row dendrogram, col dendrogram, rownames, colnames, respectively. (by default, c(5, 5, 3, 3)).</p></dd>


<dt>sepwid</dt>
<dd><p>numeric value for the width of separator between adjacent cells.</p></dd>


<dt>sepcol</dt>
<dd><p>character of color for the separator between adjacent cells</p></dd>


<dt>inner_sepcol</dt>
<dd><p>character of color for the separator between colors within 1 cell if there are more than 1.</p></dd>


<dt>save.path</dt>
<dd><p>(optional) file path to save the plot, format of the image depends on the file extension. If save.path is NULL, the heatmap will be printed. If save.path is specified, the image file will be created.</p></dd>

</dl></div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Bao Tram Vi, Sebastien Cunnac.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

