<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en"><head><meta http-equiv="Content-Type" content="text/html; charset=UTF-8"><meta charset="utf-8"><meta http-equiv="X-UA-Compatible" content="IE=edge"><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><meta name="description" content="tellTale has been primarily written to report on 'corrected' TALE RVD
sequences in indels prone, noisy DNA sequences (suboptimally polished genomes
assembly, raw reads of long read sequencing technologies [eg PacBio, ONT])
that would otherwise be missed by conventional tools (eg AnnoTALE)."><title>Search and report on the features of TALE protein domains potentially encoded
in subject DNA sequences — tellTale • tantale</title><script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no"><link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet"><script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous"><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous"><!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Search and report on the features of TALE protein domains potentially encoded
in subject DNA sequences — tellTale"><meta property="og:description" content="tellTale has been primarily written to report on 'corrected' TALE RVD
sequences in indels prone, noisy DNA sequences (suboptimally polished genomes
assembly, raw reads of long read sequencing technologies [eg PacBio, ONT])
that would otherwise be missed by conventional tools (eg AnnoTALE)."><!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]--></head><body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">tantale</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.1.9550</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto"><li class="active nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/1_tale_mining.html">1. Mining TALE sequences in genomes with tantale</a>
    <a class="dropdown-item" href="../articles/2_tale_classification.html">2. Classifying TALE sequences from genomes with tantale</a>
    <a class="dropdown-item" href="../articles/3_tale_msa.html">3. Multiple alignment of TALE sequences with tantale</a>
    <a class="dropdown-item" href="../articles/4_tale_target_prediction.html">4. TALE targets predictions with tantale</a>
    <a class="dropdown-item" href="../articles/p1_distalr.html">A comparison between `runDistal` and `distalr` implementations</a>
    <a class="dropdown-item" href="../articles/p2_multiple_alignments.html">TALEs domains sequences alignment with MAFFT</a>
    <a class="dropdown-item" href="../articles/p3_tantale_objects.html">An overview of the various objects related to TALEs in tantale</a>
  </div>
</li>
      </ul><form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off"></form>

      <ul class="navbar-nav"><li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/scunnac/tantale/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul></div>

    
  </div>
</nav><div class="container template-reference-topic">
<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Search and report on the features of TALE protein domains potentially encoded
in subject DNA sequences</h1>
      <small class="dont-index">Source: <a href="https://github.com/scunnac/tantale/blob/HEAD/R/telltale.R" class="external-link"><code>R/telltale.R</code></a></small>
      <div class="d-none name"><code>tellTale.Rd</code></div>
    </div>

    <div class="ref-description section level2">
    <p><code>tellTale</code> has been primarily written to report on 'corrected' TALE RVD
sequences in indels prone, noisy DNA sequences (suboptimally polished genomes
assembly, raw reads of long read sequencing technologies [eg PacBio, ONT])
that would otherwise be missed by conventional tools (eg AnnoTALE).</p>
    </div>

    <div class="section level2">
    <h2 id="ref-usage">Usage<a class="anchor" aria-label="anchor" href="#ref-usage"></a></h2>
    <div class="sourceCode"><pre class="sourceCode r"><code><span><span class="fu">tellTale</span><span class="op">(</span></span>
<span>  <span class="va">subjectFile</span>,</span>
<span>  outputDir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/getwd.html" class="external-link">getwd</a></span><span class="op">(</span><span class="op">)</span>,</span>
<span>  hmmFilesDir <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"hmmProfile"</span>, package <span class="op">=</span> <span class="st">"tantale"</span>, mustWork <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,</span>
<span>  TALE_NtermDNAHitMinScore <span class="op">=</span> <span class="fl">300</span>,</span>
<span>  repeatDNAHitMinScore <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  TALE_CtermDNAHitMinScore <span class="op">=</span> <span class="fl">200</span>,</span>
<span>  minDomainHitsPerSubjSeq <span class="op">=</span> <span class="fl">4</span>,</span>
<span>  mergeHits <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  minGapWidth <span class="op">=</span> <span class="fl">35</span>,</span>
<span>  appendExtremityCodes <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  rvdSep <span class="op">=</span> <span class="st">"-"</span>,</span>
<span>  hmmerpath <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"tools"</span>, <span class="st">"hmmer-3.3"</span>, <span class="st">"bin"</span>, package <span class="op">=</span> <span class="st">"tantale"</span>, mustWork <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,</span>
<span>  extendedLength <span class="op">=</span> <span class="fl">300</span>,</span>
<span>  talArrayCorrection <span class="op">=</span> <span class="cn">FALSE</span>,</span>
<span>  refForTalArrayCorrection <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/system.file.html" class="external-link">system.file</a></span><span class="op">(</span><span class="st">"extdata"</span>, <span class="st">"decipher_ref_tales_aa.fa.gz"</span>,</span>
<span>    package <span class="op">=</span> <span class="st">"tantale"</span>, mustWork <span class="op">=</span> <span class="cn">T</span><span class="op">)</span>,</span>
<span>  frameShiftCorrection <span class="op">=</span> <span class="op">-</span><span class="fl">11</span>,</span>
<span>  <span class="va">...</span></span>
<span><span class="op">)</span></span></code></pre></div>
    </div>

    <div class="section level2">
    <h2 id="arguments">Arguments<a class="anchor" aria-label="anchor" href="#arguments"></a></h2>
    <dl><dt>subjectFile</dt>
<dd><p>Fasta file with DNA sequence(s) to be searched for the
presence of TALE coding sequences (CDS).</p></dd>


<dt>outputDir</dt>
<dd><p>Path of the output directory. If not specified, results will
be written to current working folder.</p></dd>


<dt>hmmFilesDir</dt>
<dd><p>Specify the path to a folder holding the hmmfiles if you
do not want to use the ones provided with tantale.</p></dd>


<dt>TALE_NtermDNAHitMinScore</dt>
<dd><p>Minimal nhmmer score cut_off value to
consider the hit as genuine</p></dd>


<dt>repeatDNAHitMinScore</dt>
<dd><p>Minimal nhmmer score cut_off value to consider
the hit as genuine</p></dd>


<dt>TALE_CtermDNAHitMinScore</dt>
<dd><p>Minimal nhmmer score cut_off value to
consider the hit as genuine</p></dd>


<dt>minDomainHitsPerSubjSeq</dt>
<dd><p>Minimum number of nhmmer hits for a subject
sequence to be reported as having TALE diagnostic regions. This is a way to
simplify output a little by getting ride of uninformative sequences</p></dd>


<dt>mergeHits</dt>
<dd><p>Perform overlapping hits merging per domain type. Should not
be modified.</p></dd>


<dt>minGapWidth</dt>
<dd><p>Minimum gap in base pairs between two tale domain hits for
them to be considered distinct. If the length of the gap is below this
value, domains are considered "contiguous" and grouped in the same array.</p></dd>


<dt>appendExtremityCodes</dt>
<dd><p>Set this to <code>FALSE</code> if you do not want the
N- and C-TREM anchor codes in the output sequences of RVD</p></dd>


<dt>rvdSep</dt>
<dd><p>Symbol acting as a separator in RVD sequences</p></dd>


<dt>hmmerpath</dt>
<dd><p>Specify the path to a directory holding the HMMER executable
if you do not want to use the ones provided with tantale.</p></dd>


<dt>extendedLength</dt>
<dd><p>number of nucleotides to extend in 3'-end at the tal
ORF prediction stage.</p></dd>


<dt>talArrayCorrection</dt>
<dd><p>True or False</p></dd>


<dt>refForTalArrayCorrection</dt>
<dd><p>Reference AA sequences for tal array
predicted ORF correction if you do not want to use the ones provided with
tantale.</p></dd>


<dt>frameShiftCorrection</dt>
<dd><p>This is an internal parameter of the
<code><a href="https://rdrr.io/pkg/DECIPHER/man/CorrectFrameshifts.html" class="external-link">CorrectFrameshifts</a></code> function. The
default is 11 and fiddle with this at your own risk...</p></dd>


<dt>...</dt>
<dd><p>Additional parameters for the
<code><a href="https://rdrr.io/pkg/DECIPHER/man/CorrectFrameshifts.html" class="external-link">CorrectFrameshifts</a></code> function.</p></dd>

</dl></div>
    <div class="section level2">
    <h2 id="value">Value<a class="anchor" aria-label="anchor" href="#value"></a></h2>
    

<p>This functions has only side effects (writing files, mostly).
  However, if everything ran smoothly, it will invisibly return the path of
  the directory where output files were written.</p>



<p>List of output files:</p>
<p></p>
<ul><li><p>allRanges.gff: gff file of all Tal arrays detected by HMMer</p></li>
<li><p>arrayReport.tsv: report of all Tal arrays. In the arrayReport.tsv,
  column <em>predicted_dels_count</em>/<em>predicted_ins_count</em> shows the
  number of putative deletions/insertions in the raw sequences that have been
  corrected in the corrected sequences with the
  function <code><a href="https://rdrr.io/pkg/DECIPHER/man/CorrectFrameshifts.html" class="external-link">CorrectFrameshifts</a></code>.</p></li>
<li><p>hitsReport.tsv: report of all hits detected by HMMer</p></li>
<li><p>hitsReport.gff: gff file of all hits detected by HMMer</p></li>
<li><p>domainsReport.tsv: report of all Tal amino acid domains detected by AnnoTALE analyze</p></li>
<li><p>putativeTalOrf.fasta: Tal putative ORFs</p></li>
<li><p>pseudoTalCds.fasta: pseudo Tal CDS, putative Tal array ORFs detected by HMMer for whch
   AnnoTALE analyze failed to find RVD(s).</p></li>
<li><p>rvdSequences.fas: Sequence of RVDs (separated by rvdSep) predicted to be encoded in the Tal array
   ORFs by AnnoTALE. Note that if appendExtremityCodes is <code>TRUE</code> (by default),
   the N- and C-TREM anchor codes will be appended at the beginning and end of the sequences
   if the corresponding domain coding sequence was wound by HMMer at the DNA level.
   If no such HMMer hits were found, the "XXXXX" string will be appended
   to denote that AA sequences outside of the RVD array are likely to be atypical.</p></li>
<li><p>C-terminusAAAlignment.html: protein alignment of all C-termini</p></li>
<li><p>C-terminusDNAAlignment.html: DNA alignment of all C-termini</p></li>
<li><p>N-terminusAAAlignment.html: protein alignment of all N-termini</p></li>
<li><p>N-terminusDNAAlignment.html: DNA alignment of all N-termini</p></li>
<li><p>TALE_CDS_all_diagnostic_regions_hmmfile.out: HMMER profile used for tale cds search.</p></li>
<li><p>hmmerSearchOut.txt: ignore</p></li>
<li><p>nhmmerHumanReadableOutputOfLastRun.txt: primary HMMER output file.</p></li>
<li><p>tellTale.log: a log file</p></li>
<li><p>annotale folder: folder containing result of AnnoTALE analyze for all Tal arrays</p></li>
<li><p>CorrectionAlignmentAA folder: folder containing protein alignment of Tal array detected by HMMer and corrected Tal array if <code>talArrayCorrection</code> = TRUE</p></li>
<li><p>CorrectionAlignmentDNA folder: folder containing DNA alignment of Tal array detected by HMMer and corrected Tal array if <code>talArrayCorrection = TRUE</code></p></li>
</ul></div>
    <div class="section level2">
    <h2 id="details">Details<a class="anchor" aria-label="anchor" href="#details"></a></h2>
    <p>The approach is first to use <a href="http://hmmer.org/" class="external-link">HMMER</a> to find and
categorize regions in the input DNA sequence that are related to the coding
sequence of canonical TALE protein domains (N-Term, repeats, C-term). Hits
that are (nearly [see the minGapWidth parameter]) adjacent are grouped in
"taleArrays" which are considered as potential tal genes.</p>

<p>If the <code>talArrayCorrection</code> parameter is turned off, the longest
predicted open reading frame (+extendedLength) for each talArray is fed to
<a href="http://www.jstacs.de/index.php/AnnoTALE" class="external-link">AnnoTALE</a> to detect TALE
domains in the predicted translation product. The Results should hence be
very similar to what would be obtained with AnnoTALE, plus many additional
informative output files such as tabular reports.</p>
<p>If <code>talArrayCorrection</code> is turned on, these talearrays are passed to the
<code><a href="https://rdrr.io/pkg/DECIPHER/man/CorrectFrameshifts.html" class="external-link">CorrectFrameshifts</a></code> function that
attemps to 'correct' potential frameshifts in the taleArray sequences. This
conveniently removes many artefactual indels but bear in mind that this may
also <strong>erroneously</strong> 'correct' genuine frame shifts which can be highly
relevant especially for truncTALEs or iTALES. The resulting 'corrected'
taleArray open reading frames are then passed to AnnoTALE.</p>
<p>Note that occasionally, when a putative open reading frame does not encode a
canonical TALE protein (early frame shift, incomplete ORF, etc...), the
"analyze" module of AnnoTALE outputs DNA parts but no protein parts and/or
RVD sequence. This should be detected and reported in the tellTale log.</p>
    </div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside></div>


    <footer><div class="pkgdown-footer-left">
  <p></p><p>Developed by Bao Tram Vi, Sebastien Cunnac.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p><p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer></div>

  

  

  </body></html>

